tinydnssec-1.05/tinydns-data.8
.TH tinydns-data 8

.SH NAME
tinydns-data \- data tool for tinydns

.SH DESCRIPTION
This is a reference page.
For tutorial information, see the instructions for
.br
.B running a DNS server
(http://cr.yp.to/djbdns/run-server.html).

.B tinydns-data
reads local DNS information
from a file named 
.I data
in the current directory.
It creates 
.I data.cdb
in a binary format designed for
fast access by 
.BR tinydns (8).
It may also create some other files
with names beginning with 
.IR data .

.B tinydns-data
updates 
.I data.cdb
atomically,
so you can use it safely while 
.BR tinydns (8)
is running.
If anything goes wrong with the creation of 
.IR data.cdb ,
.B tinydns-data
stops and leaves the old 
.I data.cdb
in place.

.SH Data format
The DNS information in 
.I data
is a series of lines.
There are several types of lines, as shown below.

Each line starts with a special character
and continues with a series of colon-separated fields.
In some cases the fields may be omitted;
however, all colons must be included except at the end of the line.
Spaces and tabs at the end of a line are ignored.
Blank lines are ignored.

Each line contains a
.I ttl
(``time to live'')
specifying the number of seconds that the line's DNS records may be cached.
Beware that cache times below 300 seconds
will be treated as 300 by some clients,
and NS cache times below 2 seconds can cause lookup failures.
You may omit
.IR ttl ;
.B tinydns-data
will use default cache times,
carefully selected to work well in normal situations.

You may include a timestamp on each line.
If
.I ttl
is nonzero (or omitted),
the timestamp is a starting time
for the information in the line;
the line will be ignored before that time.
If 
.I ttl
is zero,
the timestamp is an ending time (``time to die'')
for the information in the line;
.BR tinydns (8)
dynamically adjusts 
.I ttl
so that the line's DNS records are not cached for more than a few seconds
past the ending time.
A timestamp is an
external TAI64 timestamp,
printed as 16 lowercase hexadecimal characters.
For example, the lines

+www.heaven.af.mil:1.2.3.4:0:4000000038af1379
+www.heaven.af.mil:1.2.3.7::4000000038af1379

specify that 
.I www.heaven.af.mil
will have address 
.I 1.2.3.4
until time 
.I 4000000038af1379
(2000-02-19 22:04:31 UTC)
and will then switch to IP address 
.IR 1.2.3.7 .

For versions 1.04 and above:
You may include a client location on each line.
The line is ignored for clients outside that location.
Client locations are specified by
.B %
lines:

%\fIlo\fR:\fIipprefix\fR

means that IP addresses starting with
.I ipprefix
are in location
.IR lo .
.I lo
is a sequence of one or two ASCII letters.
A client is in only one location;
longer prefixes override shorter prefixes.
For example,

  %in:192.168
  %ex
  +jupiter.heaven.af.mil:192.168.1.2:::in
  +jupiter.heaven.af.mil:1.2.3.4:::ex

specifies that
.I jupiter.heaven.af.mil
has address
.I 192.168.1.2
for clients in the
.I 192.168.*
network
and address
.I 1.2.3.4
for everyone else.

.SH Common data lines

.RI \. fqdn\fR:\fIip\fR:\fIx\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

Name server for our domain 
.IR fqdn .

.B tinydns-data
creates
.IP
an NS (``name server'') record
showing 
.IR x .ns.\fIfqdn\fR
as a name server for
.IR fqdn ;
.P
.IP
an A (``address'') record showing
.I ip
as the IP address
of
.IR x\fR.ns.\fIfqdn ;
and
.P
.IP
an SOA (``start of authority'') record for 
.I fqdn
listing 
.IR x\fR.ns.\fIfqdn
as the primary name server
and hostmaster@\fIfqdn\fR
as the contact address.
.P

You may have several name servers for one domain,
with a different 
.I x
for each server.

.BR tinydns (8)
will return only one SOA record per domain.

If
.I x
contains a dot
then 
.B tinydns-data
will use 
.I x
as the server name
rather than 
.IR x\fR.ns.\fIfqdn .
This feature is provided only for compatibility reasons;
names not ending with
.I fqdn
will force clients to contact parent servers
much more often than they otherwise would,
and will reduce the overall reliability of DNS.
You should omit 
.I ip
if 
.I x
has IP addresses assigned elsewhere in 
.IR data ;
in this case,
.B tinydns-data
will omit the A record.
.P
Examples:

  .panic.mil:1.8.7.55:a

creates an NS record showing
.I a.ns.panic.mil
as a name server for
.IR panic.mil ,
an A record showing
.I 1.8.7.55
as the IP address of
.IR a.ns.panic.mil ,
and an SOA record for
.IR panic.mil .

  .panic.mil:1.8.7.56:dns2.panic.mil

creates an NS record showing
.I dns2.panic.mil
as a name server for
.IR panic.mil ,
an A record showing
.I 1.8.7.56
as the IP address of
.IR dns2.panic.mil ,
and an SOA record for
.IR panic.mil .

  .panic.mil::a.ns.heaven.af.mil

creates an NS record showing
.I a.ns.heaven.af.mil
as a name server for
.IR panic.mil ,
and an SOA record for
.IR panic.mil .

.RI & fqdn\fR:\fIip\fR:\fIx\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

Name server for domain
.IR fqdn .

.B tinydns-data
creates
.IP
an NS record
showing 
.IR x\fR.ns.\fIfqdn
as a name server for 
.IR fqdn
and
.P
.IP
an A record showing
.I ip
as the IP address
of
.IR x\fR.ns.\fIfqdn .
.P

If
.I x
contains a dot
then it is treated specially; see above.

You may have several name servers for one domain,
with a different
.I x
for each server.

Normally
.I &
is used
for domains delegated by this server to child servers,
while
.I .
is used for domains delegated to this server.
.P
Examples:

  &serious.panic.mil:1.8.248.6:a

creates an NS record showing
.I a.ns.serious.panic.mil
as a name server for
.IR serious.panic.mil ,
and an A record showing
.I 1.8.248.6
as the IP address of
.IR a.ns.serious.panic.mil .

  &serious.panic.mil:1.8.248.7:ns7.panic.mil

creates an NS record showing
.I ns7.panic.mil
as a name server for
.IR serious.panic.mil ,
and an A record showing
.I 1.8.248.7
as the IP address of
.IR ns7.panic.mil .

.RI = fqdn\fR:\fIip\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

Host
.I fqdn
with IP address 
.IR ip .

.B tinydns-data
creates
.IP
an A record showing 
.I ip
as
the IP address of 
.IR fqdn
and
.P
.IP
a PTR (``pointer'') record showing 
.I fqdn
as
the name of 
.IR d.c.b.a .in-addr.arpa
if 
.I ip
is 
.IR a.b.c.d .
.P

Remember to specify name servers for some suffix of
.IR fqdn ;
otherwise 
.BR tinydns (8)
will not respond
to queries about
.IR fqdn .
The same comment applies to other records described below.
Similarly, remember to specify name servers for some suffix of
.IR d.c.b.a .in-addr.arpa,
if that domain has been delegated to you.
.P
Example:

  =button.panic.mil:1.8.7.108

creates an A record showing
.I 1.8.7.108
as the IP address of
.IR button.panic.mil ,
and a PTR record showing
.I button.panic.mil
as the name of
.IR 108.7.8.1.in-addr.arpa .

.RI + fqdn\fR:\fIip\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

Alias
.I fqdn
with IP address 
.IR ip .
This is just like 
.RI = fqdn\fR:\fIip\fR:\fIttl\fR
except that 
.B tinydns-data
does not create the PTR record.

For versions 1.04 and above:
.BR tinydns (8)
returns addresses
(from
.I +
or
.I =
or
.I @
or
.I .
or
.I &
lines)
in a random order in the answer section.
If there are more than 8 records,
it returns a random set of 8.
.P
Example:

  +button.panic.mil:1.8.7.109

creates an A record showing
.I 1.8.7.109
as another IP address for
.IR button.panic.mil .

.RI @ fqdn\fR:\fIip\fR:\fIx\fR:\fIdist\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

Mail exchanger for 
.IR fqdn .

.B tinydns-data
creates
.IP
an MX (``mail exchanger'') record
showing 
.IR x\fR.mx.\fIfqdn
as a mail exchanger for 
.I fqdn
at distance 
.I dist
and
.P
.IP
an A record showing 
.I ip
as the IP address
of 
.IR x\fR.mx.\fIfqdn .
.P
You may omit 
.IR dist ;
the default distance is 0.

If 
.I x
contains a dot
then it is treated specially; see above.

You may create several MX records for 
.IR fqdn ,
with a different 
.I x
for each server.
Make sure to arrange for the SMTP server on each IP address
to accept mail for 
.IR fqdn .
.P
Example:

  @panic.mil:1.8.7.88:mail.panic.mil

creates an MX record showing
.I mail.panic.mil
as a mail exchanger for
.I panic.mil
at distance 0, and an A record showing
.I 1.8.7.88
as the IP address of
.IR mail.panic.mil .

.RI # comment

Comment line. The line is ignored.

.SH Uncommon data lines

.RI - fqdn\fR:\fIs\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

For versions 1.04 and above:
This type of line is used by
programs that automatically edit
.B +
lines in
.I data
to temporarily exclude addresses of overloaded or dead machines.
The line is ignored.

.RI ' fqdn\fR:\fIs\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

TXT (``text'') record for 
.IR fqdn .
.B tinydns-data
creates a TXT record for 
.I fqdn
containing the string 
.IR s .
You may use octal 
.RI \ nnn
codes
to include arbitrary bytes inside 
.IR s ;
for example, 
.RI \ 072
is a colon.

.RI ^ fqdn\fR:\fIp\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

PTR record for 
.IR fqdn .
.B tinydns-data
creates a PTR record for
.I fqdn
pointing to the domain name 
.IR p .

.RI C fqdn\fR:\fIp\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

CNAME (``canonical name'') record for 
.IR fqdn .
.B tinydns-data
creates a CNAME record for 
.I fqdn
pointing to the domain name 
.IR p .

Don't use
.RI C fqdn
if there are any other records for
.I fqdn
Don't use
.RI C fqdn
for common aliases;
use
.RI + fqdn
instead.
Remember the wise words of Inigo Montoya:
``You keep using CNAME records.
I do not think they mean what you think they mean.''

.RI Z fqdn\fR:\fImname\fR:\fIrname\fR:\fIser\fR:\fIref\fR:\fIret\fR:\fIexp\fR:\fImin\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

SOA record for 
.I fqdn
showing 
.I mname
as the primary name server,
.I rname
(with the first 
.I .
converted to 
.IR @ )
as the contact address,
.I ser
as the serial number,
.I ref
as the refresh time,
.I ret
as the retry time,
.I exp
as the expire time, and
.I min
as the minimum time.
.IR ser ,
.IR ref ,
.IR ret ,
.IR exp ,
and
.I min
may be omitted;
they default to, respectively,
the modification time of the 
.I data
file,
16384 seconds,
2048 seconds,
1048576 seconds, and
2560 seconds.

.RI : fqdn\fR:\fIn\fR:\fIrdata\fR:\fIttl\fR:\fItimestamp\fR:\fIlo\fR

Generic record for
.IR fqdn .
.B tinydns-data
creates a record of type 
.I n
for 
.I fqdn
showing 
.IR rdata .
.I n
must be an integer between 1 and 65535;
it must not be 2 (NS), 5 (CNAME), 6 (SOA), 12 (PTR), 15 (MX), or 252 (AXFR).
The proper format of
.I rdata
depends on 
.IR n .
You may use octal 
.RI \ nnn
codes
to include arbitrary bytes inside 
.IR rdata .

.SH Wildcards

.I tinydns
supports wildcards of the form 
.IR *.fqdn .
Information for 
.I *.fqdn
is provided for every name ending with 
.IR .fqdn ,
.I except
names that have their own records
and names that are covered by more specific wildcards.

For example, the lines

  +pink.floyd.u.heaven.af.mil:1.2.3.4
  +*.u.heaven.af.mil:1.2.3.200

have the same effect as

  +pink.floyd.u.heaven.af.mil:1.2.3.4
  +joe.u.heaven.af.mil:1.2.3.200
  +bill.u.heaven.af.mil:1.2.3.200
  +floyd.u.heaven.af.mil:1.2.3.200
  +ishtar.u.heaven.af.mil:1.2.3.200
  +joe.bob.u.heaven.af.mil:1.2.3.200
  +sally.floyd.u.heaven.af.mil:1.2.3.200
  +post.pink.floyd.u.heaven.af.mil:1.2.3.200

and so on.

As another example, the lines

  +pink.floyd.u.heaven.af.mil:1.2.3.4
  @*.u.heaven.af.mil::mail.heaven.af.mil

have the same effect as

  +pink.floyd.u.heaven.af.mil:1.2.3.4
  @joe.u.heaven.af.mil::mail.heaven.af.mil
  @bill.u.heaven.af.mil::mail.heaven.af.mil
  @floyd.u.heaven.af.mil::mail.heaven.af.mil
  @ishtar.u.heaven.af.mil::mail.heaven.af.mil
  @joe.bob.u.heaven.af.mil::mail.heaven.af.mil
  @sally.floyd.u.heaven.af.mil::mail.heaven.af.mil
  @post.pink.floyd.u.heaven.af.mil::mail.heaven.af.mil

and so on.
Notice that the wildcard does not apply to
.IR pink.floyd.u.heaven.af.mil ,
because that name has its own records.

.SH A typical data file:

  =lion.heaven.af.mil:1.2.3.4
  @heaven.af.mil:1.2.3.4
  @3.2.1.in-addr.arpa:1.2.3.4

  =tiger.heaven.af.mil:1.2.3.5
  \.heaven.af.mil:1.2.3.5:a
  \.3.2.1.in-addr.arpa:1.2.3.5:a

  =bear.heaven.af.mil:1.2.3.6
  \.heaven.af.mil:1.2.3.6:b
  \.3.2.1.in-addr.arpa:1.2.3.6:b

  =cheetah.heaven.af.mil:1.2.3.248
  =panther.heaven.af.mil:1.2.3.249

Here is the same information in BIND zone-file format
with the two zones merged:

  heaven.af.mil. 2560 IN SOA a.ns.heaven.af.mil. hostmaster.heaven.af.mil. ...
  heaven.af.mil. 259200 IN NS a.ns.heaven.af.mil.
  heaven.af.mil. 259200 IN NS b.ns.heaven.af.mil.
  heaven.af.mil. 86400 IN MX mx.heaven.af.mil.

  3.2.1.in-addr.arpa. 2560 IN SOA a.ns.3.2.1.in-addr.arpa. hostmaster.3.2.1.in-addr.arpa. ...
  3.2.1.in-addr.arpa. 259200 IN NS a.ns.3.2.1.in-addr.arpa.
  3.2.1.in-addr.arpa. 259200 IN NS b.ns.3.2.1.in-addr.arpa.
  3.2.1.in-addr.arpa. 86400 IN MX mx.3.2.1.in-addr.arpa.

  4.3.2.1.in-addr.arpa. 86400 IN PTR lion.heaven.af.mil.
  lion.heaven.af.mil. 86400 IN A 1.2.3.4
  mx.heaven.af.mil. 86400 IN A 1.2.3.4
  mx.3.2.1.in-addr.arpa. 86400 IN A 1.2.3.4

  5.3.2.1.in-addr.arpa. 86400 IN PTR tiger.heaven.af.mil.
  tiger.heaven.af.mil. 86400 IN A 1.2.3.5
  a.ns.heaven.af.mil. 259200 IN A 1.2.3.5
  a.ns.3.2.1.in-addr.arpa. 259200 IN A 1.2.3.5

  6.3.2.1.in-addr.arpa. 86400 IN PTR bear.heaven.af.mil.
  bear.heaven.af.mil. 86400 IN A 1.2.3.6
  b.ns.heaven.af.mil. 259200 IN A 1.2.3.6
  b.ns.3.2.1.in-addr.arpa. 259200 IN A 1.2.3.6

  248.3.2.1.in-addr.arpa. 86400 IN PTR cheetah.heaven.af.mil.
  cheetah.heaven.af.mil. 86400 IN A 1.2.3.248

  249.3.2.1.in-addr.arpa. 86400 IN PTR panther.heaven.af.mil.
  panther.heaven.af.mil. 86400 IN A 1.2.3.249

.SH Design notes
The 
.I data
format is very easy for programs to edit,
and reasonably easy for humans to edit,
unlike the traditional zone-file format.

.B tinydns-data
could support a name wherever an IP address is required;
it would look up the name in DNS and use the resulting address.
This would reliably track changes in offsite IP addresses
if the database were rebuilt periodically.

.SH SEE ALSO
tinydns(8)

http://cr.yp.to/djbdns.html


tinydnssec-1.05/tinydns-data.c
#include <stdio.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "uint16.h"
#include "uint32.h"
#include "str.h"
#include "byte.h"
#include "fmt.h"
#include "ip4.h"
#include "ip6.h"
#include "exit.h"
#include "case.h"
#include "scan.h"
#include "buffer.h"
#include "strerr.h"
#include "getln.h"
#include "cdb_make.h"
#include "stralloc.h"
#include "open.h"
#include "dns.h"

#define TTL_NS 259200
#define TTL_POSITIVE 86400
#define TTL_NEGATIVE 2560

#define FATAL "tinydns-data: fatal: "

void die_datatmp(void)
{
  strerr_die2sys(111,FATAL,"unable to create data.tmp: ");
}
void nomem(void)
{
  strerr_die1sys(111,FATAL);
}

void ttdparse(stralloc *sa,char ttd[8])
{
  unsigned int i;
  char ch;

  byte_zero(ttd,8);
  for (i = 0;(i < 16) && (i < sa->len);++i) {
    ch = sa->s[i];
    if ((ch >= '0') && (ch <= '9'))
      ch -= '0';
    else if ((ch >= 'a') && (ch <= 'f'))
      ch -= 'a' - 10;
    else
      ch = 0;
    if (!(i & 1)) ch <<= 4;
    ttd[i >> 1] |= ch;
  }
}

void locparse(stralloc *sa,char loc[2])
{
  loc[0] = (sa->len > 0) ? sa->s[0] : 0;
  loc[1] = (sa->len > 1) ? sa->s[1] : 0;
}

void ipprefix_cat(stralloc *out,char *s)
{
  unsigned long u;
  char ch;
  unsigned int j;

  if (*s=='s') {
    ++s;
    if (!stralloc_catb(out,"s",1) || !stralloc_cats(out,s)) nomem();
  } else {
    if (*s=='f') ++s;
    if (!stralloc_catb(out,"f",1)) nomem();
    for (;;)
      if (*s == '.')
	++s;
      else {
	j = scan_ulong(s,&u);
	if (!j) return;
	s += j;
	ch = u;
	if (!stralloc_catb(out,&ch,1)) nomem();
      }
  }
}

void txtparse(stralloc *sa)
{
  char ch;
  unsigned int i;
  unsigned int j;

  j = 0;
  i = 0;
  while (i < sa->len) {
    ch = sa->s[i++];
    if (ch == '\\') {
      if (i >= sa->len) break;
      ch = sa->s[i++];
      if ((ch >= '0') && (ch <= '7')) {
        ch -= '0';
        if ((i < sa->len) && (sa->s[i] >= '0') && (sa->s[i] <= '7')) {
	  ch <<= 3;
	  ch += sa->s[i++] - '0';
          if ((i < sa->len) && (sa->s[i] >= '0') && (sa->s[i] <= '7')) {
	    ch <<= 3;
	    ch += sa->s[i++] - '0';
          }
        }
      }
    }
    sa->s[j++] = ch;
  }
  sa->len = j;
}

char defaultsoa[20];

void defaultsoa_init(int fd)
{
  struct stat st;
  if (fstat(fd,&st) == -1)
    strerr_die2sys(111,FATAL,"unable to stat data: ");
  uint32_pack_big(defaultsoa,st.st_mtime);
  if (byte_equal(defaultsoa,4,"\0\0\0\0"))
    defaultsoa[3] = 1;
  byte_copy(defaultsoa + 4,16,"\0\0\100\000\0\0\010\000\0\020\000\000\0\0\012\000");
}

int fdcdb;
struct cdb_make cdb;
static stralloc key;
static stralloc result;

void rr_add(const char *buf,unsigned int len)
{
  if (!stralloc_catb(&result,buf,len)) nomem();
}
void rr_addname(const char *d)
{
  rr_add(d,dns_domain_length(d));
}
void rr_start(const char type[2],unsigned long ttl,const char ttd[8],const char loc[2])
{
  char buf[4];
  if (!stralloc_copyb(&result,type,2)) nomem();
  if (byte_equal(loc,2,"\0\0"))
    rr_add("=",1);
  else {
    rr_add(">",1);
    rr_add(loc,2);
  }
  uint32_pack_big(buf,ttl);
  rr_add(buf,4);
  rr_add(ttd,8);
}
void rr_finish(const char *owner)
{
  if (byte_equal(owner,2,"\1*")) {
    owner += 2;
    result.s[2] -= 19;
  }
  if (!stralloc_copyb(&key,owner,dns_domain_length(owner))) nomem();
  case_lowerb(key.s,key.len);
  if (cdb_make_add(&cdb,key.s,key.len,result.s,result.len) == -1)
    die_datatmp();
}

buffer b;
char bspace[1024];

static stralloc line;
int match = 1;
unsigned long linenum = 0;

#define NUMFIELDS 15
static stralloc f[NUMFIELDS];

static char *d1;
static char *d2;
char dptr[DNS_NAME4_DOMAIN];
char d6ptr[DNS_NAME6_DOMAIN];

char strnum[FMT_ULONG];

void syntaxerror(const char *why)
{
  strnum[fmt_ulong(strnum,linenum)] = 0;
  strerr_die4x(111,FATAL,"unable to parse data line ",strnum,why);
}

int main()
{
  int fddata;
  int i;
  int j;
  int k;
  char ch;
  unsigned long ttl;
  char ttd[8];
  char loc[2];
  unsigned long u;
  char ip[4];
  char ip6[16];
  char type[2];
  char soa[20];
  char buf[4];

  umask(022);

  fddata = open_read("data");
  if (fddata == -1)
    strerr_die2sys(111,FATAL,"unable to open data: ");
  defaultsoa_init(fddata);

  buffer_init(&b,buffer_unixread,fddata,bspace,sizeof bspace);

  fdcdb = open_trunc("data.tmp");
  if (fdcdb == -1) die_datatmp();
  if (cdb_make_start(&cdb,fdcdb) == -1) die_datatmp();

  while (match) {
    ++linenum;
    if (getln(&b,&line,&match,'\n') == -1)
      strerr_die2sys(111,FATAL,"unable to read line: ");

    while (line.len) {
      ch = line.s[line.len - 1];
      if ((ch != ' ') && (ch != '\t') && (ch != '\n')) break;
      --line.len;
    }
    if (!line.len) continue;
    if (line.s[0] == '#') continue;
    if (line.s[0] == '-') continue;

    j = 1;
    for (i = 0;i < NUMFIELDS;++i) {
      if (j >= line.len) {
	if (!stralloc_copys(&f[i],"")) nomem();
      }
      else {
        k = byte_chr(line.s + j,line.len - j,':');
	if (!stralloc_copyb(&f[i],line.s + j,k)) nomem();
	j += k + 1;
      }
    }

    switch(line.s[0]) {

      case '%':
	locparse(&f[0],loc);
	if (!stralloc_copyb(&key,"\0%",2)) nomem();
	if (!stralloc_0(&f[1])) nomem();
	ipprefix_cat(&key,f[1].s);
        if (cdb_make_add(&cdb,key.s,key.len,loc,2) == -1)
          die_datatmp();
	break;

      case 'Z':
	if (!dns_domain_fromdot(&d1,f[0].s,f[0].len)) nomem();

	if (!stralloc_0(&f[3])) nomem();
	if (!scan_ulong(f[3].s,&u)) uint32_unpack_big(defaultsoa, (uint32 *) &u);
	uint32_pack_big(soa,u);
	if (!stralloc_0(&f[4])) nomem();
	if (!scan_ulong(f[4].s,&u)) uint32_unpack_big(defaultsoa + 4, (uint32 *) &u);
	uint32_pack_big(soa + 4,u);
	if (!stralloc_0(&f[5])) nomem();
	if (!scan_ulong(f[5].s,&u)) uint32_unpack_big(defaultsoa + 8, (uint32 *) &u);
	uint32_pack_big(soa + 8,u);
	if (!stralloc_0(&f[6])) nomem();
	if (!scan_ulong(f[6].s,&u)) uint32_unpack_big(defaultsoa + 12, (uint32 *) &u);
	uint32_pack_big(soa + 12,u);
	if (!stralloc_0(&f[7])) nomem();
	if (!scan_ulong(f[7].s,&u)) uint32_unpack_big(defaultsoa + 16, (uint32 *) &u);
	uint32_pack_big(soa + 16,u);

	if (!stralloc_0(&f[8])) nomem();
	if (!scan_ulong(f[8].s,&ttl)) ttl = TTL_NEGATIVE;
	ttdparse(&f[9],ttd);
	locparse(&f[10],loc);

	rr_start(DNS_T_SOA,ttl,ttd,loc);
	if (!dns_domain_fromdot(&d2,f[1].s,f[1].len)) nomem();
	rr_addname(d2);
	if (!dns_domain_fromdot(&d2,f[2].s,f[2].len)) nomem();
	rr_addname(d2);
	rr_add(soa,20);
	rr_finish(d1);
	break;

      case '.': case '&':
	if (!dns_domain_fromdot(&d1,f[0].s,f[0].len)) nomem();
	if (!stralloc_0(&f[3])) nomem();
	if (!scan_ulong(f[3].s,&ttl)) ttl = TTL_NS;
	ttdparse(&f[4],ttd);
	locparse(&f[5],loc);

	if (!stralloc_0(&f[1])) nomem();

	if (byte_chr(f[2].s,f[2].len,'.') >= f[2].len) {
	  if (!stralloc_cats(&f[2],".ns.")) nomem();
	  if (!stralloc_catb(&f[2],f[0].s,f[0].len)) nomem();
	}
	if (!dns_domain_fromdot(&d2,f[2].s,f[2].len)) nomem();

	if (line.s[0] == '.') {
	  rr_start(DNS_T_SOA,ttl ? TTL_NEGATIVE : 0,ttd,loc);
	  rr_addname(d2);
	  rr_add("\12hostmaster",11);
	  rr_addname(d1);
	  rr_add(defaultsoa,20);
	  rr_finish(d1);
	}

	rr_start(DNS_T_NS,ttl,ttd,loc);
	rr_addname(d2);
	rr_finish(d1);

	if (ip4_scan(f[1].s,ip)) {
	  rr_start(DNS_T_A,ttl,ttd,loc);
	  rr_add(ip,4);
	  rr_finish(d2);
	}

	break;

      case '+': case '=':
	if (!dns_domain_fromdot(&d1,f[0].s,f[0].len)) nomem();
	if (!stralloc_0(&f[2])) nomem();
	if (!scan_ulong(f[2].s,&ttl)) ttl = TTL_POSITIVE;
	ttdparse(&f[3],ttd);
	locparse(&f[4],loc);

	if (!stralloc_0(&f[1])) nomem();

	if (ip4_scan(f[1].s,ip)) {
	  rr_start(DNS_T_A,ttl,ttd,loc);
	  rr_add(ip,4);
	  rr_finish(d1);

	  if (line.s[0] == '=') {
	    dns_name4_domain(dptr,ip);
	    rr_start(DNS_T_PTR,ttl,ttd,loc);
	    rr_addname(d1);
	    rr_finish(dptr);
	  }
	}
	break;

      case '6': case '3':
	if (!dns_domain_fromdot(&d1,f[0].s,f[0].len)) nomem();
	if (!stralloc_0(&f[2])) nomem();
	if (!scan_ulong(f[2].s,&ttl)) ttl = TTL_POSITIVE;
	ttdparse(&f[3],ttd);
	locparse(&f[4],loc);

	if (!stralloc_0(&f[1])) nomem();
	if (ip6_scan_flat(f[1].s,ip6)) {
	  rr_start(DNS_T_AAAA,ttl,ttd,loc);
	  rr_add(ip6,16);
	  rr_finish(d1);

	  if (line.s[0] == '6') {	/* emit both .ip6.arpa and .ip6.int */
	    dns_name6_domain(d6ptr,ip6,DNS_IP6_ARPA);
	    rr_start(DNS_T_PTR,ttl,ttd,loc);
	    rr_addname(d1);
	    rr_finish(d6ptr);

	    dns_name6_domain(d6ptr,ip6,DNS_IP6_INT);
	    rr_start(DNS_T_PTR,ttl,ttd,loc);
	    rr_addname(d1);
	    rr_finish(d6ptr);
	  }
	}
	break;

      case '@':
	if (!dns_domain_fromdot(&d1,f[0].s,f[0].len)) nomem();
	if (!stralloc_0(&f[4])) nomem();
	if (!scan_ulong(f[4].s,&ttl)) ttl = TTL_POSITIVE;
	ttdparse(&f[5],ttd);
	locparse(&f[6],loc);

	if (!stralloc_0(&f[1])) nomem();

	if (byte_chr(f[2].s,f[2].len,'.') >= f[2].len) {
	  if (!stralloc_cats(&f[2],".mx.")) nomem();
	  if (!stralloc_catb(&f[2],f[0].s,f[0].len)) nomem();
	}
	if (!dns_domain_fromdot(&d2,f[2].s,f[2].len)) nomem();

	if (!stralloc_0(&f[3])) nomem();
	if (!scan_ulong(f[3].s,&u)) u = 0;

	rr_start(DNS_T_MX,ttl,ttd,loc);
	uint16_pack_big(buf,u);
	rr_add(buf,2);
	rr_addname(d2);
	rr_finish(d1);

	if (ip4_scan(f[1].s,ip)) {
	  rr_start(DNS_T_A,ttl,ttd,loc);
	  rr_add(ip,4);
	  rr_finish(d2);
	}
	break;

      case '^': case 'C':
	if (!dns_domain_fromdot(&d1,f[0].s,f[0].len)) nomem();
	if (!dns_domain_fromdot(&d2,f[1].s,f[1].len)) nomem();
	if (!stralloc_0(&f[2])) nomem();
	if (!scan_ulong(f[2].s,&ttl)) ttl = TTL_POSITIVE;
	ttdparse(&f[3],ttd);
	locparse(&f[4],loc);

	if (line.s[0] == 'C')
	  rr_start(DNS_T_CNAME,ttl,ttd,loc);
	else
	  rr_start(DNS_T_PTR,ttl,ttd,loc);
	rr_addname(d2);
	rr_finish(d1);
	break;

      case '\'':
	if (!dns_domain_fromdot(&d1,f[0].s,f[0].len)) nomem();
	if (!stralloc_0(&f[2])) nomem();
	if (!scan_ulong(f[2].s,&ttl)) ttl = TTL_POSITIVE;
	ttdparse(&f[3],ttd);
	locparse(&f[4],loc);

	rr_start(DNS_T_TXT,ttl,ttd,loc);

	txtparse(&f[1]);
	i = 0;
	while (i < f[1].len) {
	  k = f[1].len - i;
	  if (k > 255) k = 255;
	  ch = k;
	  rr_add(&ch,1);
	  rr_add(f[1].s + i,k);
	  i += k;
	}

	rr_finish(d1);
	break;

      case ':':
	if (!dns_domain_fromdot(&d1,f[0].s,f[0].len)) nomem();
	if (!stralloc_0(&f[3])) nomem();
	if (!scan_ulong(f[3].s,&ttl)) ttl = TTL_POSITIVE;
	ttdparse(&f[4],ttd);
	locparse(&f[5],loc);

	if (!stralloc_0(&f[1])) nomem();
	scan_ulong(f[1].s,&u);
	uint16_pack_big(type,u);
	if (byte_equal(type,2,DNS_T_AXFR))
	  syntaxerror(": type AXFR prohibited");
	if (byte_equal(type,2,"\0\0"))
	  syntaxerror(": type 0 prohibited");
	if (byte_equal(type,2,DNS_T_SOA))
	  syntaxerror(": type SOA prohibited");
	if (byte_equal(type,2,DNS_T_NS))
	  syntaxerror(": type NS prohibited");
	if (byte_equal(type,2,DNS_T_CNAME))
	  syntaxerror(": type CNAME prohibited");
	if (byte_equal(type,2,DNS_T_PTR))
	  syntaxerror(": type PTR prohibited");
	if (byte_equal(type,2,DNS_T_MX))
	  syntaxerror(": type MX prohibited");

	txtparse(&f[2]);

	rr_start(type,ttl,ttd,loc);
	rr_add(f[2].s,f[2].len);
	rr_finish(d1);
	break;

      default:
        syntaxerror(": unrecognized leading character");
    }
  }

  if (cdb_make_finish(&cdb) == -1) die_datatmp();
  if (fsync(fdcdb) == -1) die_datatmp();
  if (close(fdcdb) == -1) die_datatmp(); /* NFS stupidity */
  if (rename("data.tmp","data.cdb") == -1)
    strerr_die2sys(111,FATAL,"unable to move data.tmp to data.cdb: ");

  _exit(0);
}
